processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -147,23 +135,14 @@ protected Integer addToFacts(
             makeColumnSelectorFactory(agg, rowSupplier, deserializeComplexMetrics)
         );
       }
-      final Integer rowIndex = indexIncrement.getAndIncrement();
-
-      concurrentSet(rowIndex, aggs);
-
-      // Last ditch sanity checks
-      if (numEntries.get() >= maxRowCount && !facts.containsKey(key)) {
+      if (numEntries.incrementAndGet() > maxRowCount) {
+        numEntries.decrementAndGet();
```

Comment:
I think we can safely stick with the old logic and only increment numEntries once we put the aggregators in the map